packages/snaps-controllers/src/snaps/endowments/name-lookup.ts

```diff
@@ -0,0 +1,147 @@
+import type {
+  EndowmentGetterParams,
+  PermissionSpecificationBuilder,
+  PermissionValidatorConstraint,
+  ValidPermissionSpecification,
+  Caveat,
+  CaveatSpecificationConstraint,
+  PermissionConstraint,
+} from '@metamask/permission-controller';
+import { PermissionType, SubjectType } from '@metamask/permission-controller';
+import { SnapCaveatType } from '@metamask/snaps-utils';
+import type { Json, NonEmptyArray } from '@metamask/utils';
+import { assert, hasProperty, isPlainObject } from '@metamask/utils';
+import { ethErrors } from 'eth-rpc-errors';
+
+import { SnapEndowments } from './enum';
+
+const permissionName = SnapEndowments.NameLookup;
+
+type NameLookupEndowmentSpecification = ValidPermissionSpecification<{
+  permissionType: PermissionType.Endowment;
+  targetName: typeof permissionName;
+  endowmentGetter: (_options?: EndowmentGetterParams) => undefined;
+  allowedCaveats: Readonly<NonEmptyArray<string>> | null;
+  validator: PermissionValidatorConstraint;
+}>;
+
+/**
+ * `endowment:name-lookup` returns nothing; it is intended to be used as a flag
+ * by the extension to detect whether the snap has the capability to resolve a domain/address.
+ *
+ * @param _builderOptions - Optional specification builder options.
+ * @returns The specification for the name-lookup endowment.
+ */
+const specificationBuilder: PermissionSpecificationBuilder<
+  PermissionType.Endowment,
+  any,
+  NameLookupEndowmentSpecification
+> = (_builderOptions?: unknown) => {
+  return {
+    permissionType: PermissionType.Endowment,
+    targetName: permissionName,
+    allowedCaveats: [SnapCaveatType.ChainIds],
+    endowmentGetter: (_getterOptions?: EndowmentGetterParams) => undefined,
+    validator: ({ caveats }) => {
+      if (
+        !caveats ||
+        (caveats !== null && caveats?.length > 1) ||
+        (caveats?.length === 1 && caveats[0].type !== SnapCaveatType.ChainIds)
+      ) {
+        throw ethErrors.rpc.invalidParams({
+          message: `Expected a single "${SnapCaveatType.ChainIds}" caveat.`,
+        });
+      }
+    },
+    subjectTypes: [SubjectType.Snap],
+  };
+};
+
+export const nameLookupEndowmentBuilder = Object.freeze({
+  targetName: permissionName,
+  specificationBuilder,
+} as const);
+
+/**
+ * Validates the type of the caveat value.
+ *
+ * @param caveat - The caveat to validate.
+ * @throws If the caveat value is invalid.
+ */
+function validateCaveat(caveat: Caveat<string, any>): void {
+  if (!hasProperty(caveat, 'value') || !isPlainObject(caveat)) {
+    throw ethErrors.rpc.invalidParams({
+      message: 'Expected a plain object.',
+    });
+  }
+
+  const { value } = caveat;
+
+  assert(
+    Array.isArray(value) && value.every((val) => typeof val === 'string'),
```

Comment:
Should we validate that they are CAIP2 chain IDs?

Reply:
Good point, yes.